design/ceph/object/object-store-backend.md

```diff
@@ -0,0 +1,103 @@
+transformations of operations (such as redirection to different backends), or caching layers that
+cut across multiple backends.
+
+The primary Store is the current RadosStore, which provides access to a Ceph cluster via RADOS.
+However, new Stores are possible that store the data in any desired platform. One such Store, called
+DBStore, has been developed that stores data in SQL, and specifically in a local SQLite database.
+
+## Goals
+
+Rook-Ceph has a tremendous support for the Ceph Object Store and we want to leverage that.
+Especially the various controllers that are reconciling ObjectStore, Multisite configurations and
+Bucket provisioning.
+Mainly we are looking at providing the next generation of RGW via Rook.
+
+# Proposal details
+
+The `CephObjectStore` API will have a few details added to its spec such as an image so that it can
+be deployed without a Ceph cluster.
+
+```go
+type ObjectStoreSpec struct {
+...
+	// Image is the name of the image to use to deploy the object store
+	Image string `json:"image"`
+...
+}
+```
+
+Also, we need to expose a new type for the backend to use:
+
+```go
+type GatewaySpec struct {
+...
+	// Backend refers to the backend to use for the object store
+	// +optional
+	Backend *ObjectStoreBackend `json:"backend,omitempty"`
+...
+}
+```
+
+Where the `ObjectStoreBackend` is represented as such:
+
+```go
+type ObjectStoreBackend struct {
+	// The name of the backend
+	// +optional
+	// +kubebuilder:validation:Enum={"rados","sqlite"}
+	Name string `json:"name,omitempty"`
+
+    // Config is the configuration for the backend
+	// +optional
+    Config map[string]string `json:"config,omitempty"`
+}
+```
+
+Then the `CephObjectStore` `spec.go` file will need to be adapted depending on the backend.
+
+## Risks and Mitigation
+
+We still need to determine what the best approach between using the Rook-Ceph operator as is or
+building a new operator.
+
+Some pros to use the Rook-Ceph operator:
+* It is easier to use
+* Not much code changes involved
+* No need to change the various controller packages
+* When COSI is available we can integrate and deploy it
+* Code is maintained at a single location
+
+Some cons to use the Rook-Ceph operator:
+* More resources (controllers) are started than needed
+* The name Rook-Ceph implies that Ceph is being deployed which can be confusing if a "standalone"
+  version of RGW is being used.
+
+Some pros to use a new Operator:
+* It's much lighter
+* The base image most likely does not need Ceph packages installed so we can have minimal image. One
+  technique for this is to forward all the radosgw commands (only one to create the RGW admin Op
+  user) to the RGW pod directly.
+
+Some cons to use a new Operator:
+* New Operator to write
+* Use and import object/lib-bucket-prov controllers
+* Refactor a little bit object/lib-bucket-prov controllers packages to avoid using internal struct
```

Comment:
I assume we won't focus on lib-bucket-prov with this project.

Reply:
I guess @leseb referred to the controller in Rook which uses a lib-bucket provisioner than the actual library.

Reply:
Right, @thotz is correct, just the lib-bucket-prov controller not actual package.